packages/models-library/src/models_library/user_preferences.py

```diff
@@ -0,0 +1,177 @@
+        return new_class
+
+
+PreferenceName: TypeAlias = str
+PreferenceIdentifier: TypeAlias = str
+
+
+class _ExtendedBaseModel(BaseModel, metaclass=_AutoRegisterMeta):
+    ...
+
+
+class PreferenceType(StrAutoEnum):
+    BACKEND = auto()
+    FRONTEND = auto()
+    USER_SERVICE = auto()
+
+
+class NoPreferenceFoundError(RuntimeError):
+    ...
+
+
+class _BaseUserPreferenceModel(_ExtendedBaseModel):
+    preference_type: PreferenceType = Field(
+        ..., description="distinguish between the types of preferences"
+    )
+
+    value: Any = Field(
+        ...,
+        description="the value of the preference. Stored as is and cannot be queried over",
+    )
+
+    @classmethod
+    def get_preference_class_from_name(
+        cls, preference_name: PreferenceName
+    ) -> type["_BaseUserPreferenceModel"]:
+        preference_class: type[
+            "_BaseUserPreferenceModel"
+        ] | None = cls._registered_user_preference_classes.get(
+            preference_name, None
+        )  # type: ignore
+        if preference_class is None:
+            msg = f"No preference class found for provided {preference_name=}"
+            raise NoPreferenceFoundError(msg)
+        return preference_class
+
+    @classmethod
+    def get_preference_name(cls) -> PreferenceName:
+        # NOTE: this will be `unique` among all subclasses.
+        # No class inherited from this one, can be defined using the same name,
+        # even if the context is different.
+        return cls.__name__
+
+    @classmethod
+    def get_default_value(cls) -> Any:
+        return (
+            cls.__fields__["value"].default_factory()
+            if cls.__fields__["value"].default_factory
+            else cls.__fields__["value"].default
+        )
+
+    def dict(  # noqa: A003
+        self,
+        *,
+        include: AbstractSetIntStr | MappingIntStrAny = None,  # type: ignore
+        exclude: AbstractSetIntStr | MappingIntStrAny = None,  # type: ignore
+        by_alias: bool = False,
+        skip_defaults: bool = None,  # type: ignore  # noqa: RUF013
+        exclude_unset: bool = False,
+        exclude_defaults: bool = False,
+        exclude_none: bool = False,
+    ) -> DictStrAny:
+        if exclude is None:
+            config_class = getattr(self, "Config", None)
+            exclude_from_serialization: set | None = getattr(
+                config_class, "exclude_from_serialization", None
+            )
+            if exclude_from_serialization:
+                exclude = exclude_from_serialization
+        return super().dict(
+            include=include,
+            exclude=exclude,
+            by_alias=by_alias,
+            skip_defaults=skip_defaults,
+            exclude_unset=exclude_unset,
+            exclude_defaults=exclude_defaults,
+            exclude_none=exclude_none,
+        )
+
+
+class BaseBackendUserPreference(_BaseUserPreferenceModel):
+    preference_type: PreferenceType = Field(PreferenceType.BACKEND, frozen=True)
```

Comment:
so you are using a discriminator here correct?
please look in here how Pydantic recommands how to do it:
https://docs.pydantic.dev/latest/usage/types/unions/#discriminated-unions-aka-tagged-unions

Using const=True allows to set it as unchangeable default